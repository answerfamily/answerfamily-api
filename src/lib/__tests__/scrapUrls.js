jest.mock('graphql-request');
const { request } = require('graphql-request');

const MockDate = require('mockdate');

const {
  loadMongoFixtures,
  unloadMongoFixtures,
} = require('../../../test/fixtures');
const fixtures = require('../__fixtures__/scrapUrls');
const scrapUrls = require('../scrapUrls');
const DataLoaders = require('../../dataloaders');
const client = require('../mongoClient');

describe('scrapping & storage', () => {
  let server;

  afterAll(async () => {
    await client.deleteByQuery({
      index: 'urls',
      body: {
        query: {
          match_all: {},
        },
      },
    });
    server.close();
  });

  it(
    'scraps from Internet and handles error',
    async () => {
      MockDate.set(1485593157011);

      request.mockReturnValue(
        Promise.resolve({
          data: {
            resolvedUrls: [
              {
                url: 'http://example.com/index.html',
                canonical: 'http://example.com/index.html',
                title: 'Some title',
                summary: 'Some text as summary',
                topImageUrl: '',
                html: '<html><head></head><body>Hello world</body></html>',
                status: 200,
              },
              {
                url: 'http://example.com/not-found',
                canonical: 'http://example.com/not-found',
                title: '',
                summary: 'Not Found',
                topImageUrl: '',
                html: '<html><head></head><body>Not Found</body></html>',
                status: 404,
              },
            ],
          },
        })
      );

      const [foundResult, notFoundResult] = await scrapUrls(
        `
          This should work: http://example.com/index.html
          This should be not found: http://example.com/not-found
          This should not match: http://malformedUrl:100000
        `,
        { client }
      );
      MockDate.reset();

      expect(gql.__getRequests()).toMatchSnapshot('GraphQL requests');
      gql.__reset();

      expect(foundResult).toMatchSnapshot('foundResult');
      expect(notFoundResult).toMatchSnapshot('notFoundResult');

      // scrapUrls() don't wait until refresh, thus refresh on our own
      await client.indices.refresh({ index: 'urls' });

      const {
        hits: { hits: urls },
      } = await client.search({
        index: 'urls',
        body: {
          query: {
            match_all: {},
          },
        },
      });

      expect(urls.map(({ _source }) => _source)).toMatchSnapshot(
        'Docs stored to urls index'
      );
    },
    30000
  );
});

describe('caching', () => {
  beforeAll(() => loadFixtures(fixtures));
  afterAll(() => unloadFixtures(fixtures));

  it('fetches latest fetch from urls', async () => {
    const loaders = new DataLoaders();
    expect(
      await scrapUrls(
        `
          http://example.com/article/1111-aaaaa-bbb-ccc // full URL
          http://example.com/article/1111 // canonical URL
        `,
        {
          cacheLoader: loaders.urlLoader,
          noFetch: true,
        }
      )
    ).toMatchSnapshot();
  });
});
