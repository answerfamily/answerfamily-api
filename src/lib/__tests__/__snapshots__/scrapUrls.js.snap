// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`caching fetches latest fetch from urls 1`] = `
Array [
  Object {
    "_id": "5c4dc54fe36f8525bf6f6399",
    "canonical": "http://example.com/article/1111",
    "fetchedAt": 2017-02-01T00:00:00.000Z,
    "fromUrlsCache": true,
    "status": 200,
    "summary": "Changed summary",
    "title": "Changed title",
    "url": "http://example.com/article/1111-aaaaa-bbb-ccc",
  },
  null,
]
`;

exports[`scrapping & storage scraps from Internet and handles error: Docs stored to urls index 1`] = `
Array [
  Object {
    "canonical": "http://example.com/index.html",
    "fetchedAt": 2017-01-28T08:45:57.011Z,
    "html": "<html><head></head><body>Hello world</body></html>",
    "status": 200,
    "summary": "Some text as summary",
    "title": "Some title",
    "topImageUrl": "",
    "url": "http://example.com/index.html",
  },
  Object {
    "canonical": "http://example.com/not-found",
    "fetchedAt": 2017-01-28T08:45:57.011Z,
    "html": "<html><head></head><body>Not Found</body></html>",
    "status": 404,
    "summary": "Not Found",
    "title": "",
    "topImageUrl": "",
    "url": "http://example.com/not-found",
  },
]
`;

exports[`scrapping & storage scraps from Internet and handles error: GraphQL requests 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "http://localhost:4000",
      "
        query($urls: [String]!) {
          resolvedUrls(urls: $urls) {
            url
            canonical
            title
            summary
            topImageUrl
            status
            error
          }
        }
      ",
      Object {
        "urls": Array [
          "http://example.com/index.html",
          "http://example.com/not-found",
        ],
      },
    ],
  ],
  "results": Array [
    Object {
      "isThrow": false,
      "value": Promise {},
    },
  ],
}
`;

exports[`scrapping & storage scraps from Internet and handles error: foundResult 1`] = `
Object {
  "canonical": "http://example.com/index.html",
  "html": "<html><head></head><body>Hello world</body></html>",
  "status": 200,
  "summary": "Some text as summary",
  "title": "Some title",
  "topImageUrl": "",
  "url": "http://example.com/index.html",
}
`;

exports[`scrapping & storage scraps from Internet and handles error: notFoundResult 1`] = `
Object {
  "canonical": "http://example.com/not-found",
  "html": "<html><head></head><body>Not Found</body></html>",
  "status": 404,
  "summary": "Not Found",
  "title": "",
  "topImageUrl": "",
  "url": "http://example.com/not-found",
}
`;
